README
			     What is XRN?

  XRN is a Usenet News reader for the X window system.  It requires an
NNTP server to talk to (i.e., it cannot read news from local disk
files), and the NNTP server must support the XHDR command (most
servers nowadays do; if yours doesn't, XRN will complain when you try
to run it).

  Versions of XRN prior to 7.00 were maintained by Ellen Santovich and
Rick Spickelmier.  XRN is currently maintained by Jonathan Kamens.
But reports or comments about it should be sent to bug-xrn@cam.ov.com.
Please do *not* write to that address with basic questions about
compiling the program; if you are unfamiliar with compiling X
applications and the instructions below are not sufficient, you should
seek help at your site.


			   Where to get XRN

  The most recent production release of XRN can always be retrieved
from <URL:ftp://ftp.cam.ov.com/pub/xrn/xrn.tgz>.  The most recent
beta-test release of XRN (if there is a beta release more recent than
the last production release) can always be retrieved from
<URL:ftp://ftp.cam.ov.com/pub/xrn/beta.tgz>.  the file
<URL:ftp://ftp.cam.ov.com/pub/xrn/xrn-linux.tgz> contains a binary XRN
release for Linux, based on the current production or beta release,
whichever is more recent.  These files are GNU tar archives compressed
with GNU zip.

  The current production release and the current Linux binary release
are also available in the X Consortium's contributed software archive,
in the directory <ftp://ftp.x.org/contrib/applications/xrn>.


	      Interesting files in the XRN distribution

  There are a number of files in the XRN distribution besides the
actual source files for the program.  I maintain some of these files
actively.  Others are left over from before I started maintaining XRN,
and I have not reviewed them for correctness or updated them; use the
information in them at your own risk.  In the following list,
asterisks indicate files which I have tried to keep up-to-date (this
more files may be updated in future releases, especially if people
send me feedback about the out-of-date ones):

COMMON-PROBLMS		A list of common problems and questions about
			the program, with explanations and solutions.
COPYRIGHT		Describes the copyright restrictions on XRN.
			I have left the University of California
			copyright intact, despite the fact that it's
			not clear at this point how much of the code
			was actually written by UC employees, because
			that's the easiest thing to do and it doesn't
			do any harm.
* ChangeLog		A list of changes from previous releases of
			XRN.  Note that because the last release of
			XRN was so long ago, and because the new
			release is so different from the previous
			release, I have not included ChangeLog entries
			from previous releases. I.e., the ChangeLog
			starts with release 7.00-beta-1.

			Starting with release 7.01-beta-3, significant
			user-visible changes are marked with an
			asterisk in the second column so that they can
			be spotted easily.
* README		This file.
* TODO			A list of pending bug fixes and improvements
			to XRN.  If you're interested in helping to
			develop XRN, picking one of the items in this
			list and digging around in the source code to
			see what you can do about it is a good way to
			start.

			This file is sort of "half up-to-date,"
			because I've been adding new items to the
			beginning of it, but the items at the end of
			it are left over from before I took over
			maintenance of XRN, and some of them may be
			inaccurate.
Xresources.sam		A sample of how you might customize XRN in
			your own X resources.
contrib/*		Various scripts, hints, etc. contributed by
			XRN users and not "officially" part of XRN.
* doc/rfc977.txt	The Internet RFC governing NNTP, the Network
			News Transport Protocol, which XRN users to
			talk to the server.
* doc/strtok.doc	Documentation of the strtok() implementation
			included with XRN.
* doc/vprintf.doc	Documentation for the vprintf() implementation
			included with XRN.
* doc/vprintf.port	The original vprintf() implementation from
			which the one used in XRN is derived.
still-to-do/*		Various bug fixes that have not yet been
			examined or incorporated into the program.
			Some of these are probably obsolete.  I
			inherited them when I took over maintenance
			of the program, and I haven't done anything
			with them.


		      Supported/Tested Platforms

  XRN has been compiled and run on a large variety of platforms; it
should compile pretty easily on most platforms.  The maintainer uses
SunOS 4.1.3, gcc 2.4.5, and X11R5 libraries.  Other successful
platforms reported by users include OSF/1, HP-UX, AIX, Solaris 2.x,
BSD/386, Linux, and SGI.

  Rick Murphy <murphy@ufp.dco.dec.com> has a version of XRN that runs
on VMS, available as <ftp://gatekeeper.dec.com/pub/DEC/dxrn.tar.Z>.
His version of XRN also will compile with a DECwindows interface on
UNIX, as well as with a Motif interface with some "bells and whistles"
that this version of XRN does not have.  Integrating DXRN into the
code base of XRN (i.e., this package) is on my TODO list, but I doubt
it'll get done in the short term.


		     Compiling and installing XRN

  The following steps are necessary in order to compile and install
XRN.  The last few steps are the standard method for compiling and
installing an X program; if you need more help with them, you should
look for it at your site.

1. Edit the Imakefile (after making it writeable with chmod) for your
   local configuration.  The comments in the file explain what needs
   to be edited.  Pay particular attention to "INN", which enables and
   support for InterNetNews (INN).

2. Edit config.h (after making it writeable with chmod) for your local
   configuration.  The file is (sort of) arranged from most
   significant to least significant parameters, so if you get part-way
   through it and start to feel like you can just leave the rest
   as-is, you probably can.

   If you don't define a parameter that needs to be defined, you'll
   get an error during compilation.

   Make sure to delete the "#error" line near the top of config.h when
   you're done editing it for your local configuration.  That line is
   there to prevent you from trying to compile the program without
   configuring it first.  Do NOT define OV_CAMBRIDGE when compiling
   the program -- I use that symbol for my local configuration so that
   I can ship the same config.h as the one I use, so if you define it,
   you'll end up getting the configuration that I use locally (which
   is certainly not what you want).

   If you want to put all of your configuration paramaters in the
   SITE_DEFINES variable in the Imakefile, you can do that instead of
   editing config.h.  However, if you do that, make sure to put
   "-DCONFIG_H_IS_OK" in SITE_DEFINES.

3. Run "xmkmf" and then "make".

   If you don't have xmkmf at your site, or if it doesn't work or
   produces a Makefile that doesn't work, you can try configuring by
   hand, by editing the Makefile included in the distribution (note
   that if you ran xmkmf, it probably renamed the distributed Makefile
   to Makefile.bak).  It's generated using xmkmf at my site, and
   almost certainly will require editing to be correct for yours.  If
   you have to do this and you have trouble, please get help from
   someone at your site rather than contacting me about it; I don't
   have time to support sites that don't have xmkmf installed
   properly.

   If you get an error about a missing header file "mesg_strings.h"
   when you run "make", then run "make mesg_strings.h" and then run
   "make" again.

4. Run "make install" to install the program and its application
   defaults file, and/or "make install.man" to install the man page.

   Note that XRN will not run correctly without its application
   defaults file installed.


			 The XRN mailing list

  If you would like to be informed of new releases of XRN and of any
major bugs that are discovered in between releases, you can subscribe
to the xrn-users@cam.ov.com mailing list.  It's an announcement list,
not a discussion list, so the traffic on it is very low.  If you're
interested, send mail to xrn-users-request@cam.ov.com and ask to be
added to the list.



I hope you find XRN useful!

  Jonathan Kamens
  jik@cam.ov.com

# $Id: README,v 1.39 1995-01-31 16:31:12 jik Exp $


ChangeLog
Tue Sep  5 19:10:21 1995  Jonathan Kamens  <jik@annex-1-slip-jik.cam.ov.com>

	* XRn.src, patchlevel.h, xrn-man.src: Update version number for
 	7.04-beta-3.

Tue Sep  5 18:57:22 1995  Jonathan Kamens  <jik@annex-1-slip-jik.cam.ov.com>

 *	* compose.c, compose.h, mesg_strings.c, ngMode.c: Don't let the
 	user quit or exit from the program if a composition is pending.

	* artMode.c: If killing an author instead of a subject, then the
 	message displayed after killing should say "Author has been
 	killed" instead of "Subject has been killed".

	* server.c (getlineslist): Don't retrieve anything if
 	displayLineCount is false.

	* xthelper.c, xthelper.h: Add support in xthWaitForMapped for
 	waiting for an expose event too.

	* xthelper.c: Add some debugging support to
 	xthHandleAllPendingEvents.

	* xrn.c: xthWaitForMapped() now has an extra argument to specify
 	whether an expose event should also be waited for.  We're not
 	using that functionality here, but we still need to specify False
 	as the new argument.

	* internals.c, internals.h: Make it possible for fillUpArray() to
 	be aborted part-way through.

	* cursor.c: fillUpArray() now has another argument and a return
 	value, because it can be told to allow an abort in the middle of
 	its work.  We're not using that functionality here, but we still
 	need to account for the new argument.

 *	* cancel.c: When creating a cancel button, wait until it's mapped
 	and gets an expose event, and then process all pending events so
 	that its label will be drawn properly, before returning.  This
 	way, cancel buttons will be popped up and drawn when they're
 	supposed to be instead of being delayed.

 *	* artMode.c: Add a cancel button to "List Old" in article mode.
  	This was suggested by George Ross <gdmr@dcs.ed.ac.uk>.
  	Unfortunately, it's not possible to do this really well without
 	significantly slowing down the performance of the "List Old"
 	command, at least not until the program is multi-threaded or talks
 	to the NNTP server in a sub-process or something like that.  With
 	this change, the "List Old" can be aborted either one-third of the
 	way through the work it does or two-thirds of the way through.
  	It's not very good granularity, but it's better than nothing.

	* resources.c: Alphabetize the options in the usage message, and
 	delete one redundant one.

Tue Aug 22 09:34:20 1995  Jonathan Kamens  <jik@annex-1-slip-jik.cam.ov.com>

	* compose.c: Another fix for the GENERATE_EXTRA_FIELDS Date code
 	from nissen@montefiore.ulg.ac.be (Alain Nissen).

 *	* internals.c: Fix bug reported by Michael Schmitz
 	<schmitz@ph-cip.Uni-Koeln.DE>: "Users that do not have a .newsrc
 	file will not see the NEWUSER_GROUPS after startup. I experienced
 	that this is independent of the platform xrn runs on. Added
 	newsgroups will appear in the ng mode (if they are not empty) but
 	the NEWUSER_GROUPS will only appear after doing a 'rescan'.  On
 	the other hand, when you leave without doing a rescan, all
 	articles in the NEWUSER_GROUPS are marked as read!"

	* config.h: "Vyacheslav Shubinsky, Computer Specialist, Towson, MD
 	"<slava@usgs.gov> says that we need to use putenv on DGUX.

	* config.h: There were two blocks of code defining USE_PUTENV.  I
 	removed one of them.

	* server.c (getlist): Verify that each line returned by the XHDR
 	command is in the proper format; if not, display an error message
 	and ignore it.  This is an attempt to fix or get more information
 	about a problem reported by Joe Konstan <konstan@cs.umn.edu>.

 *	* compose.c, mesg_strings.c: Improvement suggested by Harald
 	Leinders <hl@arthur.ph2.Uni-Koeln.DE>: when the signature file is
 	too big to be included, display a warning instead of ignoring it
 	silently, so that the user knows why it was ignored (instead of
 	thinking that XRN is broken).

Fri Jul  7 16:47:52 1995  Jonathan Kamens  <jik@annex-1-slip-jik.cam.ov.com>

	* resources.c (Initialize): Fix a small memory leak.

Thu Jun 29 19:03:50 1995  Jonathan Kamens  <jik@annex-1-slip-jik.cam.ov.com>

	* compose.c (includeHandler): Close a file included in an article
 	when done including it.  Bug pointed out by
 	philippe@pablo.physics.lsa.umich.edu (Philippe Brieu).

	* mesg.c (_info): Keep track of which widget we actually last set
 	a message in, as well as what the message was, so that if the info
 	widget changes, we know to update it.

Fri Jun 16 12:36:31 1995  Jonathan Kamens  <jik@annex-1-slip-jik.cam.ov.com>

	* server.c: Don't free the filename until after using it to
 	generate an error in getarticle().

Mon Jun 12 19:23:51 1995  Jonathan Kamens  <jik@annex-1-slip-jik.cam.ov.com>

	* COMMON-PROBLMS, README, README.Linux (new file): Document that
 	there are problems with the regular-expression routines in Linux
 	libc.

Fri Jun  9 22:08:21 1995  Jonathan Kamens  <jik@annex-1-slip-jik.cam.ov.com>

	* utils.c: Add parentheses around an expression used as a truth
 	value.  Pointed out by Jean-Marc.Lasgouttes@inria.fr (Jean-Marc
 	Lasgouttes).

	* internals.c: Move '{' and '}' into the list of characters which
 	should be quoted in regexps by enclosing them in square braces,
 	rather than by backslashing them.

Thu Jun  8 02:05:07 1995  Jonathan Kamens  <jik@annex-1-slip-jik.cam.ov.com>

	* artMode.c, buttons.c, config.h, internals.c, internals.h,
 	server.c: Support POSIX regular expression routines, and use them
 	on Linux.

Wed Jun  7 01:29:00 1995  Jonathan Kamens  <jik@annex-1-slip-jik.cam.ov.com>

	* server.c: Patch from "David A. Curry" <davy@ecn.purdue.edu> to
 	fix a null-pointer dereference when parsing "From:" lines when
 	authorFullName is false.

Tue Jun  6 23:49:50 1995  Jonathan Kamens  <jik@annex-1-slip-jik.cam.ov.com>

 *	* ngMode.c: Fix bug reported by
 	philippe@pablo.physics.lsa.umich.edu (Philippe Brieu): Newsgroup
 	mode says "No more unread articles in the subscribed to
 	newsgroups" even when there are, in fact, unread articles.


xrn.c
#if !defined(lint) && !defined(SABER) && !defined(GCC_WALL)
static char XRNrcsid[] = "$Id: xrn.c,v 1.36 2005-12-01 08:49:34 jik Exp $";
#endif

/*
 * xrn - an X-based NNTP news reader
 *
 * Copyright (c) 1988-1993, Ellen M. Sentovich and Rick L. Spickelmier.
 *
 * Permission to use, copy, modify, and distribute this software and its
 * documentation for any purpose and without fee is hereby granted, provided
 * that the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of the University of California not
 * be used in advertising or publicity pertaining to distribution of 
 * the software without specific, written prior permission.  The University
 * of California makes no representations about the suitability of this
 * software for any purpose.  It is provided "as is" without express or
 * implied warranty.
 *
 * THE UNIVERSITY OF CALIFORNIA DISCLAIMS ALL WARRANTIES WITH REGARD TO 
 * THIS SOFTWARE, INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND 
 * FITNESS, IN NO EVENT SHALL THE UNIVERSITY OF CALIFORNIA BE LIABLE FOR
 * ANY SPECIAL, INDIRECT OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER
 * RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF
 * CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN 
 * CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
 * xrn.c: set up the main screens
 *
 */

#include "copyright.h"
#include "config.h"
#include "utils.h"
#include <X11/Xos.h>
#include <X11/Intrinsic.h> /* so we have Widget */
#ifdef MOTIF
# include <Xm/Xm.h>
#endif

#include "news.h"
#include "xthelper.h"
#include "xmisc.h"
#include "resources.h"
#include "internals.h"
#include "error_hnds.h"
#include "buttons.h"
#include "mesg.h"
#include "xrn.h"
#include "compose.h"
#include "mesg_strings.h"
#include "InfoLine.h"
#include "Frame.h"
#include "ngMode.h"
#include "file_cache.h"

#ifdef XFILESEARCHPATH
static void AddPathToSearchPath _ARGUMENTS((char *));
#endif

/* global variables that represent the widgets that are dynamically changed */

Widget TopLevel;
XtAppContext TopContext;
Widget TopInfoLine;      /* top button info line                      */
Widget BottomInfoLine;   /* bottom button info line                   */

int XRNState;            /* XRN status: news and x                    */

int inchannel, outchannel;
file_cache FileCache;

/*ARGSUSED*/
int main(argc, argv)
    int argc;
    char **argv;
{
    Widget frame;

    int sv[2];

    pipe (sv);
    inchannel = sv[0];
    outchannel = sv[1];

    XRNState = 0;

#ifdef XFILESEARCHPATH
    AddPathToSearchPath(XFILESEARCHPATH);
#endif
    
    TopLevel = Initialize(argc, argv);

    if (app_resources.cacheFilesMaxFiles < 10)
      app_resources.cacheFilesMaxFiles = 10;
    if (app_resources.cacheFilesMaxSize < 0)
      app_resources.cacheFilesMaxSize = 1;
    
    ehInstallSignalHandlers();
    ehInstallErrorHandlers();

    FileCache = file_cache_create(app_resources.tmpDir, "xrn",
				  app_resources.cacheFilesMaxFiles,
				  app_resources.cacheFilesMaxSize);

    if (app_resources.geometry != NIL(char)) {
       GetMainFrameSize(TopLevel, (app_resources.geometry));
    }
    
    /* create the pane and its widgets */

    frame = CreateMainFrame(TopLevel);

    TopInfoLine = InfoLineCreate("info", "", frame);
    BottomInfoLine = 0;
    
    createButtons();

    /* create the icon */
    xmIconCreate();

    XrnAddCloseCallbacks(TopLevel);

    /*
      Be sure that initializeNews() and determineMode() get called after
      the main window is already realized.  -- jik 11/13/94
      */
    
    XtRealizeWidget(TopLevel);
    XRNState |= XRN_X_UP;
    xthWaitForMapped(TopLevel, False);
#ifdef MOTIF
    XmUpdateDisplay(TopLevel);
#endif

    /* initialize the news system, read the newsrc file */
    initializeNews();
    XRNState |= XRN_NEWS_UP;

    /* set up the text window, mode buttons, and question */
    determineMode(True);

    xrnUnbusyCursor();

    if (app_resources.version == 0) {
	mesgPane(XRN_SERIOUS, 0, NO_APP_DEFAULTS_MSG);
    } else if (strcmp(app_resources.version, PACKAGE_VERSION) != 0) {
	mesgPane(XRN_SERIOUS, 0, NO_APP_DEFAULTS_MSG);
	mesgPane(XRN_SERIOUS | XRN_APPEND, 0, VERSIONS_MSG, app_resources.version,
		 PACKAGE_VERSION);
    }

    XtAppAddInput(TopContext,
                inchannel, (XtPointer) XtInputReadMask, processMessage, (XtPointer) 0);

#if XtSpecificationRelease < 6
    MyMainLoop(TopContext);
#else
    XtAppMainLoop(TopContext);
#endif
    exit(0);
}       

#ifdef XFILESEARCHPATH
static void AddPathToSearchPath(path)
    char *path;
{
     char *old, *new;

     old = getenv("XFILESEARCHPATH");
     if (old) {
#ifdef USE_PUTENV
	  /* +1 for =, +2 for :, +3 for null */
	  new = XtMalloc((Cardinal) (strlen("XFILESEARCHPATH") +
				     strlen(old) +
				     strlen(path) + 3));
	  (void) strcpy(new, "XFILESEARCHPATH");
	  (void) strcat(new, "=");
	  (void) strcat(new, old);
	  (void) strcat(new, ":");
	  (void) strcat(new, path);
	  putenv(new);
#else
	  /* +1 for colon, +2 for null */
	  new = XtMalloc((Cardinal) (strlen(old) + strlen(path) + 2));
	  (void) strcpy(new, old);
	  (void) strcat(new, ":");
	  (void) strcat(new, path);
	  setenv("XFILESEARCHPATH", new, 1);
#endif
     }
     else {
#ifdef USE_PUTENV
	  new = XtMalloc((Cardinal) (strlen("XFILESEARCHPATH") +
				     strlen(path) + 2));
	  (void) strcpy(new, "XFILESEARCHPATH");
	  (void) strcat(new, "=");
	  (void) strcat(new, path);
	  putenv(new);
#else
	  setenv("XFILESEARCHPATH", path, 1);
#endif
     }
}
#endif

#if XtSpecificationRelease < 6
static XEvent last_event;

XEvent *XtLastEventProcessed(display)
     Display *display;
{
  return &last_event;
}

void MyMainLoop(app)
     XtAppContext app;
{
  XEvent event;

  for (;;) {
    XtAppNextEvent(app, &event);
    MyDispatchEvent(&event);
  }
}

Boolean MyDispatchEvent(event)
     XEvent *event;
{
  last_event = *event;
  return XtDispatchEvent(event);
}
#endif /* XtSpecificationRelease < 6 */
